staticaddr/loopin/actions.go

```diff
@@ -0,0 +1,594 @@
+			Memo:       "static address loop-in",
+			Expiry:     3600 * 24 * 365,
+			RouteHints: f.loopIn.RouteHints,
+		},
+	)
+	if err != nil {
+		return f.HandleError(err)
+	}
+	f.loopIn.SwapInvoice = swapInvoice
+
+	f.loopIn.ProtocolVersion = version.AddressProtocolVersion(
+		version.CurrentRPCProtocolVersion(),
+	)
+
+	loopInReq := &looprpc.ServerStaticAddressLoopInRequest{
+		SwapHash:         f.loopIn.SwapHash[:],
+		DepositOutpoints: f.loopIn.DepositOutpoints,
+		HtlcClientPubKey: f.loopIn.ClientPubkey.SerializeCompressed(),
+		SwapInvoice:      f.loopIn.SwapInvoice,
+		ProtocolVersion:  version.CurrentRPCProtocolVersion(),
+		UserAgent:        loop.UserAgent(f.loopIn.Initiator),
+	}
+	if f.loopIn.LastHop != nil {
+		loopInReq.LastHop = f.loopIn.LastHop
+	}
+
+	loopInResp, err := f.cfg.StaticAddressServerClient.ServerStaticAddressLoopIn( //nolint:lll
+		f.ctx, loopInReq,
+	)
+	if err != nil {
+		return f.HandleError(err)
+	}
+
+	serverPubkey, err := btcec.ParsePubKey(loopInResp.HtlcServerPubKey)
+	if err != nil {
+		return f.HandleError(err)
+	}
+	f.loopIn.ServerPubkey = serverPubkey
+
+	// Validate if the response parameters are outside our allowed range
+	// preventing us from continuing with a swap.
+	err = f.cfg.ValidateLoopInContract(
+		int32(f.loopIn.InitiationHeight), loopInResp.HtlcExpiry,
+	)
+	if err != nil {
+		return f.HandleError(err)
+	}
+
+	f.loopIn.HtlcCltvExpiry = loopInResp.HtlcExpiry
+	f.htlcServerNonces, err = toNonces(loopInResp.HtlcServerNonces)
+	if err != nil {
+		return f.HandleError(err)
+	}
+
+	// We need to defend against the server setting high fees for the htlc
+	// tx since we might have to sweep the timeout path. We maximally allow
+	// 20% of the swap value to be spent on fees.
+	f.loopIn.HtlcTxFeeRate = chainfee.SatPerKWeight(loopInResp.HtlcFeeRate)
+	fee := f.loopIn.HtlcTxFeeRate.FeeForWeight(f.loopIn.htlcWeight())
+
+	if fee > f.loopIn.TotalDepositAmount()/5 {
+		return f.HandleError(errors.New("fee is higher than 20% of " +
+			"sweep value"))
+	}
+
+	// Derive the sweep address for the htlc timeout sweep tx.
+	sweepAddress, err := f.cfg.WalletKit.NextAddr(
+		f.ctx, lnwallet.DefaultAccountName,
+		walletrpc.AddressType_TAPROOT_PUBKEY, false,
+	)
+	if err != nil {
+		return f.HandleError(err)
+	}
+	f.loopIn.HtlcTimeoutSweepAddress = sweepAddress
+
+	// Once the htlc tx is initiated, we store the loop-in in the database.
+	err = f.cfg.Store.CreateLoopIn(f.ctx, f.loopIn)
+	if err != nil {
+		return f.HandleError(err)
+	}
+
+	return OnHtlcInitiated
+}
+
+// SignHtlcTxAction is called if the htlc was initialized and the server
+// provided the necessary information to construct the htlc tx. We sign the htlc
+// tx and send the signatures to the server.
+func (f *FSM) SignHtlcTxAction(_ fsm.EventContext) fsm.EventType {
+	var err error
+
+	f.loopIn.AddressParams, err = f.cfg.AddressManager.GetStaticAddressParameters( //nolint:lll
+		f.ctx,
+	)
+	if err != nil {
+		return f.HandleError(err)
```

Comment:
nit: Generally, slightly more logging could be helpful as `HandleError` is too generic.

Reply:
added more context to errors in the file.